### PR TITLE
Webservice - Enable key by hosts name or IP

### LINKED ISSUE #21493: Webservice - Enable key by hosts name or IP
## Is your feature request related to a problem?

Improve the security of Web Services. In some cases a key is not enough to secure access to the API. It would be interesting to have the possibility of limiting accesses by IP address or Domain name.

See : https://devdocs.prestashop.com/1.7/webservice/getting-started/
Capture du 2020-10-19 00-49-15 (1)

## Describe the solution you'd like

- Ability to enable or disable host or IP check.
- When verification is enabled, only declared IPs and hosts can access the API for this key.

image

## Specs
**Form**
A toggle "**Host check**" is displayed at the end of the form. By default, the toggle is disabled. 
If enabled, a text area "**Host allowed**" is displayed to enter all the allowed IPs. They must be split by a comma. 

If the user has enabled the "**Host check**" and saves the form without entering any IP, then an error is thrown, the user stays on the current page and the saving is canceled. 

After been saved when the user edits the web service key from the list, all the information (list of IPS allowed and the saved configuration) must be displayed.

If the text area contains URLs and the user disables the "**Host check**", on the reactivation of the "**Host check**" the old URLs are displayed again in the text area. (Bonus, if it can't be done: the text area is cleared each time it is deactivated)

**List**
The list displays the "Check Host" status (enabled or disabled). It can't be edited from the list, since it has to add the URL to save the changes.

**Missing wording or wording to be challenged:**
- Host check is it enough explicit
- small text bellow the toggle "Host check"
- small text bellow the text is "Host allowed" if needed

### PR DESCRIPTION
**Pull request by PululuK**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Improve the security of Web Services. In some cases a key is not enough to secure access to the API. It would be interesting to have the possibility of limiting accesses by IP address or Domain name.
| Type?         | improvement 
| Category?     | WS 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/21493
| How to test?  | See https://github.com/PrestaShop/PrestaShop/issues/21493

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21494)

### CODE REVIEW

File: classes/webservice/WebserviceKey.php [hunk 1]
```diff
@@ -183,4 +191,24 @@ public static function setPermissionForAccount($id_account, $permissions_to_set)
 
         return $ok;
     }
+
+    /**
+     * @param string $auth_key
+     *
+     * @return WebserviceKey|null
+     */
+    public static function getWebServiceObjectFromKey($auth_key)
```
Comment (Progi1984) [inline classes/webservice/WebserviceKey.php]:
```suggestion
    public static function getWebServiceObjectFromKey(string $auth_key): ?WebserviceKey
```

File: classes/webservice/WebserviceRequest.php [hunk 2]
```diff
@@ -1887,4 +1891,28 @@ public static function getallheaders()
 
         return $retarr;
     }
+
+    /**
+     * Check if client hostname/ip is allowed.
+     *
+     * @return bool
+     */
+    protected function checkHostAllowed($key)
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
    protected function checkHostAllowed(string $key): bool
```

File: classes/webservice/WebserviceRequest.php [hunk 3]
```diff
@@ -1887,4 +1891,28 @@ public static function getallheaders()
 
         return $retarr;
     }
+
+    /**
+     * Check if client hostname/ip is allowed.
+     *
+     * @return bool
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
     * Check if client hostname/ip is allowed.
     *
     * @param string $key
     *
     * @return bool
```

File: src/Core/Domain/Webservice/Command/AddWebserviceKeyCommand.php [hunk 4]
```diff
@@ -118,4 +134,20 @@ public function getAssociatedShops()
     {
         return $this->associatedShops;
     }
+
+    /**
+     * @return string
+     */
+    public function getHostsAllowed()
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/AddWebserviceKeyCommand.php]:
```suggestion
    public function getHostsAllowed(): string
```

File: src/Core/Domain/Webservice/Command/AddWebserviceKeyCommand.php [hunk 5]
```diff
@@ -118,4 +134,20 @@ public function getAssociatedShops()
...
+    {
+        return $this->hosts_allowed;
+    }
+
+    /**
+     * @return bool
+     */
+    public function getHostsCheck()
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/AddWebserviceKeyCommand.php]:
```suggestion
    public function getHostsCheck(): bool
```

File: src/Adapter/Webservice/CommandHandler/EditWebserviceKeyHandler.php [hunk 6]
```diff
@@ -87,6 +87,9 @@ private function updateLegacyWebserviceKeyWithCommandData(
             $webserviceKey->active = $command->getStatus();
         }
 
+        $webserviceKey->hosts_allowed = $command->getHostsAllowed();
+        $webserviceKey->hosts_check = (bool) $command->getHostsCheck();
```
Comment (Progi1984) [inline src/Adapter/Webservice/CommandHandler/EditWebserviceKeyHandler.php]:
```suggestion
        $webserviceKey->hosts_check = $command->getHostsCheck();
```

File: src/Adapter/Webservice/CommandHandler/AddWebserviceKeyHandler.php [hunk 7]
```diff
@@ -80,6 +80,8 @@ private function createLegacyWebserviceKeyFromCommand(AddWebserviceKeyCommand $c
         $webserviceKey->key = $command->getKey()->getValue();
         $webserviceKey->description = $command->getDescription();
         $webserviceKey->active = $command->getStatus();
+        $webserviceKey->hosts_allowed = $command->getHostsAllowed();
+        $webserviceKey->hosts_check = (bool) $command->getHostsCheck();
```
Comment (Progi1984) [inline src/Adapter/Webservice/CommandHandler/AddWebserviceKeyHandler.php]:
```suggestion
        $webserviceKey->hosts_check = $command->getHostsCheck();
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 8]
```diff
@@ -179,4 +189,44 @@ public function setShopAssociation(array $shopAssociation)
 
         return $this;
     }
+
+    /**
+     * @return string|null
+     */
+    public function getHostsAllowed()
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php:196]:
```suggestion
    public function getHostsAllowed(): ?string
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 9]
```diff
@@ -179,4 +189,44 @@ public function setShopAssociation(array $shopAssociation)
...
+    {
+        return $this->hosts_allowed;
+    }
+
+    /**
+     * @param string $hosts_allowed
+     *
+     * @return self
+     */
+    public function setHostsAllowed($hosts_allowed)
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php]:
```suggestion
    public function setHostsAllowed(string $hosts_allowed): self
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 10]
```diff
@@ -179,4 +189,44 @@ public function setShopAssociation(array $shopAssociation)
...
+    {
+        $this->hosts_allowed = $hosts_allowed;
+
+        return $this;
+    }
+
+    /**
+     * @return bool|null
+     */
+    public function getHostsCheck()
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php:216]:
```suggestion
    public function getHostsCheck(): ?bool
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 11]
```diff
@@ -179,4 +189,44 @@ public function setShopAssociation(array $shopAssociation)
...
+    {
+        return $this->hosts_check;
+    }
+
+    /**
+     * @param bool $hosts_check
+     *
+     * @return self
+     */
+    public function setHostsCheck($hosts_check)
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php]:
```suggestion
    public function setHostsCheck(bool $hosts_check): self
```

File: src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php [hunk 12]
```diff
@@ -63,28 +63,44 @@ class EditableWebserviceKey
...
      */
     public function __construct(
         WebserviceKeyId $webserviceKeyId,
         $key,
         $description,
         $status,
         array $resourcePermissions,
-        array $associatedShops
+        array $associatedShops,
+        string $hosts_allowed = null,
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php]:
```suggestion
        string $hosts_allowed = '',
```

File: src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php [hunk 13]
```diff
@@ -134,4 +150,20 @@ public function getAssociatedShops()
     {
         return $this->associatedShops;
     }
+
+    /**
+     * @return string|null
+     */
+    public function getHostsAllowed()
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php]:
```suggestion
    public function getHostsAllowed(): string
```

File: src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php [hunk 14]
```diff
@@ -134,4 +150,20 @@ public function getAssociatedShops()
     {
         return $this->associatedShops;
     }
+
+    /**
+     * @return string|null
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php]:
```suggestion
     * @return string
```

File: src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php [hunk 15]
```diff
@@ -134,4 +150,20 @@ public function getAssociatedShops()
...
+     */
+    public function getHostsAllowed()
+    {
+        return $this->hosts_allowed;
+    }
+
+    /**
+     * @return bool|null
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php]:
```suggestion
     * @return bool
```

File: src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php [hunk 16]
```diff
@@ -134,4 +150,20 @@ public function getAssociatedShops()
...
+     */
+    public function getHostsCheck()
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/QueryResult/EditableWebserviceKey.php]:
```suggestion
    public function getHostsCheck(): bool
```

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Webservice/WebserviceKeyType.php [hunk 17]
```diff
@@ -113,6 +113,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'scrollable' => false,
                 'headers_to_disable' => ['all'],
             ])
+            ->add('hosts_allowed', TextareaType::class, [
+                'required' => false,
+                'empty_data' => '',
+                'attr' => [
+                    'rows' => 5,
+                    'placeholder' => 'www.prestashop.com, 42.24.4.2,127.0.0.1,99.98.97.96',
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Webservice/WebserviceKeyType.php]:
```suggestion
                    'placeholder' => 'www.prestashop.com,42.24.4.2,127.0.0.1,99.98.97.96',
```

File: src/Core/Domain/Webservice/Command/AddWebserviceKeyCommand.php [hunk 18]
```diff
@@ -58,25 +58,41 @@ class AddWebserviceKeyCommand
...
+     * @param bool $hosts_check
      */
     public function __construct(
         $key,
         $description,
         $status,
         array $permissions,
-        array $associatedShops
+        array $associatedShops,
+        string $hosts_allowed = null,
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/AddWebserviceKeyCommand.php]:
```suggestion
        string $hosts_allowed = '',
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 19]
```diff
@@ -64,6 +64,16 @@ class EditWebserviceKeyCommand
...
 
+    /**
+     * @var string|null
+     */
+    private $hosts_allowed;
+
+    /**
+     * @var bool|null
+     */
+    private $hosts_check;
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php]:
```suggestion
    /**
     * @var string
     */
    private $hosts_allowed = '';

    /**
     * @var bool
     */
    private $hosts_check = false;
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 20]
```diff
@@ -179,4 +189,44 @@ public function setShopAssociation(array $shopAssociation)
...
         return $this;
     }
+
+    /**
+     * @return string|null
+     */
+    public function getHostsAllowed()
+    {
+        return $this->hosts_allowed;
+    }
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php]:
```suggestion
    /**
     * @return string
     */
    public function getHostsAllowed(): string
    {
        return $this->hosts_allowed;
    }
```

File: src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php [hunk 21]
```diff
@@ -179,4 +189,44 @@ public function setShopAssociation(array $shopAssociation)
...
+        return $this;
+    }
+
+    /**
+     * @return bool|null
+     */
+    public function getHostsCheck()
+    {
+        return $this->hosts_check;
+    }
```
Comment (Progi1984) [inline src/Core/Domain/Webservice/Command/EditWebserviceKeyCommand.php]:
```suggestion
    /**
     * @return bool
     */
    public function getHostsCheck(): bool
    {
        return $this->hosts_check;
    }
```

**Review by Progi1984 — COMMENTED** [review]
Some Type Hints

**Comment by PululuK** [comment]
> Some Type Hints

Thank you very much @Progi1984

**Review by Progi1984 — DISMISSED** [review]

**Comment by Julievrz** [comment]
Hello @PululuK 😄 Are you done with the wording of this PR? I saw on the issue that it was said that it could be good to have a wording displayed above "Enable hosts check" to explain the feature. I would be happy to help you with this but I would need more information as it is a more technical feature...

**Comment by kpodemski** [comment]
Hey @PululuK it's still `WIP`?

**Comment by PululuK** [comment]
> Hey @PululuK it's still `WIP`?

Yes ! Sorry I forgot to add WIP tag

**Comment by PululuK** [comment]
> Hey @PululuK it's still `WIP`?

@kpodemski ready for reviews 🧐

File: classes/webservice/WebserviceRequest.php [hunk 22]
```diff
@@ -812,6 +812,10 @@ protected function authenticate()
                     if (!$this->keyPermissions) {
                         $this->setError(401, 'No permission for this authentication key', 21);
                     }
+
+                    if (!$this->checkHostAllowed($this->_key, Tools::getRemoteAddr())) {
+                        $this->setError(503, 'No permission for this host authentication', 666);
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php:817]: :imp:

File: src/Adapter/Webservice/WebserviceKeyHostsCheckModifier.php [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Webservice;
```
Comment (Progi1984) [inline src/Adapter/Webservice/WebserviceKeyHostsCheckModifier.php]:
```suggestion
 */
 
declare(strict_types=1);

namespace PrestaShop\PrestaShop\Adapter\Webservice;
```

**Review by Progi1984 — COMMENTED** [review]
A small feedback but it's ok

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Discussion in progress in the issue.

**Comment by Julievrz** [comment]
@MatShir Is this PR closed too?

**Comment by PululuK** [comment]
> @MatShir Is this PR closed too?

Yes @Julievrz

**Pull request closed without merging**